## Introduction to Data Science

#### University of Redlands - DATA 101
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data101.joannabieri.com](https://joannabieri.com/data101.html)

---------------------------------------
# Homework Day 11
---------------------------------------

GOALS:

1. Understand Causal vs Experimental Studies
2. Do a more free form data analysis
3. Start doing your ethics reading

----------------------------------------------------------

This homework has **2 questions**, **10 exercises** and **1 essay response**.

In [3]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

from itables import show

# This stops a few warning messages from showing
pd.options.mode.chained_assignment = None 
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

In [4]:
file_name = 'data/yougov-climate.csv'
DF = pd.read_csv(file_name,index_col=0)
DF

The climate is changing and human activity is mainly responsible    \
country                                                                       
GB                                                     833                    
US                                                     507                    

         The climate is changing and human activity is partly responsible, together with other factors    \
country                                                                                                    
GB                                                     604                                                 
US                                                     493                                                 

         The climate is changing but human activity is not responsible at all    \
country                                                                           
GB                                                      49                        
US                                                     120                        

         The climate is not changing    Don't know  
country                                             
GB                                  33         114  
US                                  80         133

In the lecture we analyzed the following questions for the first column in the data:

### First question:

What percent of **all respondents** think the climate is changing and human activity is mainly responsible?  

### Second question:

What percent of **GB respondents** think the climate is changing and  human activity is mainly responsible?  

### Third Question

What percent of **US respondents** think the climate is changing and  human activity is mainly responsible?  

**Q1** Repeat the analysis from the lecture notes for one of the other columns/questions.

1. Percent total
2. Percent from GB
3. Percent from US

Talk about the conditional probability in this case:

eg. In a person is from ________ then there is a ________ probability that they believe _______. If a person answered _______ then they are more likley to be from ________. 

In [9]:
DF.loc['total']= DF.sum(axis=0)
DF['total'] = DF.sum(axis=1)
DF

The climate is changing and human activity is mainly responsible    \
country                                                                       
GB                                                     833                    
US                                                     507                    
total                                                 1340                    

         The climate is changing and human activity is partly responsible, together with other factors    \
country                                                                                                    
GB                                                     604                                                 
US                                                     493                                                 
total                                                 1097                                                 

         The climate is changing but human activity is not responsible at all    \
country                                                                           
GB                                                      49                        
US                                                     120                        
total                                                  169                        

         The climate is not changing    Don't know  total  
country                                                    
GB                                  33         114   1633  
US                                  80         133   1333  
total                              113         247   2966

In [11]:
all_respondents = DF['total'].loc['total']
human_responsible = DF['The climate is changing and human activity is mainly responsible  ']['total']
human_responsible/all_respondents

0.45178691840863117

In [13]:
gb_respondants = DF['total'].loc['GB']
GB_human_responsible = DF['The climate is changing and human activity is mainly responsible  ']['GB']
GB_human_responsible/gb_respondants

0.5101041028781383

In [15]:
us_respondants = DF['total'].loc['US']
US_human_responsible = DF['The climate is changing and human activity is mainly responsible  ']['US']
US_human_responsible/us_respondants

0.3803450862715679

First Question:
45% of all respondents think the climate is changing and human activity is mainly responsible.

SeconQ question51%ent fritish GB respondents think the climate is changing and human activity is mainly responsi.l
?

Third Que:s38%ercent of US respondents think the climate is changing and human activity is mainly respo.

Q1
Someone from the US has a 38% probability that they believe climate change is mainly due to human activity. If a person answered they do believe climate change is mainly caused by human activity then they are more likley to be from Britain.sible?

## Berkeley admission data example

- Study carried out by the Graduate Division of the University of California, Berkeley in the early 70’s to evaluate whether there was a gender bias in graduate admissions.
- The data come from six departments. For confidentiality we'll call them A-F. 
- We have information on whether the applicant was male or female and whether they were admitted or rejected. This is an old study so only two binary classifications were used. 

### Here is the data


In [24]:
file_name = 'data/berkley.csv'
DF = pd.read_csv(file_name)
show(DF)

In [26]:
DF_melt = pd.melt(DF,id_vars=['Department'],var_name='MF',value_name='Number')
DF_melt

Department          MF  Number
0           A    Male Yes     512
1           B    Male Yes     353
2           C    Male Yes     120
3           D    Male Yes     138
4           E    Male Yes      53
5           F    Male Yes      22
6           A     Male No     313
7           B     Male No     207
8           C     Male No     205
9           D     Male No     279
10          E     Male No     138
11          F     Male No     351
12          A  Female Yes      89
13          B  Female Yes      17
14          C  Female Yes     202
15          D  Female Yes     131
16          E  Female Yes      94
17          F  Female Yes      24
18          A   Female No      19
19          B   Female No       8
20          C   Female No     391
21          D   Female No     244
22          E   Female No     299
23          F   Female No     317

In [28]:
DF_melt['gender'] = DF_melt['MF'].apply(lambda x: x.split(' ')[0]).copy()
DF_melt['admitted'] = DF_melt['MF'].apply(lambda x: x.split(' ')[1]).copy()

In [30]:
DF_melt

Department          MF  Number  gender admitted
0           A    Male Yes     512    Male      Yes
1           B    Male Yes     353    Male      Yes
2           C    Male Yes     120    Male      Yes
3           D    Male Yes     138    Male      Yes
4           E    Male Yes      53    Male      Yes
5           F    Male Yes      22    Male      Yes
6           A     Male No     313    Male       No
7           B     Male No     207    Male       No
8           C     Male No     205    Male       No
9           D     Male No     279    Male       No
10          E     Male No     138    Male       No
11          F     Male No     351    Male       No
12          A  Female Yes      89  Female      Yes
13          B  Female Yes      17  Female      Yes
14          C  Female Yes     202  Female      Yes
15          D  Female Yes     131  Female      Yes
16          E  Female Yes      94  Female      Yes
17          F  Female Yes      24  Female      Yes
18          A   Female No      19  Female       No
19          B   Female No       8  Female       No
20          C   Female No     391  Female       No
21          D   Female No     244  Female       No
22          E   Female No     299  Female       No
23          F   Female No     317  Female       No

In [39]:
my_columns = ['Number','gender','admitted']
DF_totals = DF_melt[my_columns].groupby(['gender','admitted'],as_index=False).sum()
DF_totals

gender admitted  Number
0  Female       No    1278
1  Female      Yes     557
2    Male       No    1493
3    Male      Yes    1198

In [41]:
mask = DF_totals['gender'] == 'Male'
num_male = DF_totals[mask]['Number'].sum()
mask = DF_totals['gender'] == 'Female'
num_female = DF_totals[mask]['Number'].sum()

mask = (DF_totals['gender'] == 'Male') & (DF_totals['admitted'] == 'Yes')
prob_male = DF_totals[mask]['Number']/num_male
mask = (DF_totals['gender'] == 'Female') & (DF_totals['admitted'] == 'Yes')
prob_female = DF_totals[mask]['Number']/num_female

In [43]:
prob_male

3    0.445188
Name: Number, dtype: float64

In [45]:
prob_female

1    0.303542
Name: Number, dtype: float64

### Gender distribution by department

What can we say about the gender distribution if we look at the individual departments.

Start with our original "melted" data frame:

#### Lets pivot!

Pivot this data so that our departments become the column labels and our MF column becomes the index.

In [47]:
DF_dept = DF_melt.pivot(index='MF',columns='Department',values='Number')
DF_dept

Department    A    B    C    D    E    F
MF                                      
Female No    19    8  391  244  299  317
Female Yes   89   17  202  131   94   24
Male No     313  207  205  279  138  351
Male Yes    512  353  120  138   53   22

In [49]:
prob_female = DF_dept['A']['Female Yes']/(DF_dept['A']['Female Yes']+DF_dept['A']['Female No'])
prob_male = DF_dept['A']['Male Yes']/(DF_dept['A']['Male Yes']+DF_dept['A']['Male No'])

In [51]:
prob_female

0.8240740740740741

In [53]:
prob_male

0.6206060606060606

**Q2** Calculate the proportions for the other departments. We did department A in the lecture. You can use the code from the lecture or come up with your own way to do this. You can do this one by one or try using a FOR loop.

Talk about what you see in these proportions. Which departments admit a large number of women? How do your number align with the visualization below?

In [56]:
prob_female = DF_dept['B']['Female Yes']/(DF_dept['B']['Female Yes']+DF_dept['B']['Female No'])
prob_male = DF_dept['B']['Male Yes']/(DF_dept['B']['Male Yes']+DF_dept['B']['Male No'])

In [58]:
prob_female

0.68

In [60]:
prob_male

0.6303571428571428

In [62]:
prob_female = DF_dept['C']['Female Yes']/(DF_dept['C']['Female Yes']+DF_dept['C']['Female No'])
prob_male = DF_dept['C']['Male Yes']/(DF_dept['C']['Male Yes']+DF_dept['C']['Male No'])

In [64]:
prob_female

0.3406408094435076

In [66]:
prob_male

0.36923076923076925

In [68]:
prob_female = DF_dept['D']['Female Yes']/(DF_dept['D']['Female Yes']+DF_dept['D']['Female No'])
prob_male = DF_dept['D']['Male Yes']/(DF_dept['D']['Male Yes']+DF_dept['D']['Male No'])

In [70]:
prob_female

0.34933333333333333

In [72]:
prob_male

0.33093525179856115

In [74]:
prob_female = DF_dept['E']['Female Yes']/(DF_dept['E']['Female Yes']+DF_dept['E']['Female No'])
prob_male = DF_dept['E']['Male Yes']/(DF_dept['E']['Male Yes']+DF_dept['E']['Male No'])

In [80]:
prob_female

0.23918575063613232

In [82]:
prob_male

0.2774869109947644

In [84]:
prob_female = DF_dept['F']['Female Yes']/(DF_dept['F']['Female Yes']+DF_dept['F']['Female No'])
prob_male = DF_dept['F']['Male Yes']/(DF_dept['F']['Male Yes']+DF_dept['F']['Male No'])

In [86]:
prob_female

0.07038123167155426

In [88]:
prob_male

0.058981233243967826

**Q2**
- Departments A and B admit the most women out of all of the departments
- Females tend to have a higher probability of admittance for all besides C and E
- these numbers don't seem to correlate completely with the graph below.

### Here is a plot of the proportions data

In [23]:
fig = px.histogram(DF_melt,
                   y='gender',
                   x='Number',
                   barnorm = "percent",
                   color='admitted',
                   facet_col='Department',
                   facet_col_wrap=2,
                   color_discrete_map = {'No':'#99d8c9','Yes':'#2ca25f'})

fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[1]))
fig.update_xaxes(title_text='')

fig.update_layout(title='Percent male and female applications',
                  title_x=0.5,
                  template="plotly_white",
                  xaxis_title="",
                  yaxis_title="Gender",
                  legend_title='Admitted',
                  autosize=False,
                  width=800,
                  height=500)

fig.show()

------------------------
------------------------

## Homework - Exercises

** Homework content and data from Data Science in a Box - he-05-legos.Rmd

This week we'll do some data gymnastics to refresh and review what we learned over the past few weeks using (**simulated**) data from Lego sales in 2018 for a sample of customers who bought Legos in the US. This is different than the data we used on the Exam!


### Data

In [92]:
file_name = 'data/lego-sales.csv'
DF = pd.read_csv(file_name)
DF

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
615     Talise   Nieukirk   16  801-555-2343   24902    41556   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
615                  Mixels              Series 7  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
615                                  Tiketz    62.0      4.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
615  http://images.brickset.com/sets/images/41556-1...         2  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[620 rows x 14 columns]

### Exercises

* Answer the following questions using reproducible Python code.
* For each question, state your answer in a sentence, e.g. "In this sample, the first three common names of purchasers are ...".
* Note that the answers to all questions are within the context of this particular sample of sales, i.e. you shouldn't make inferences about the population of all Lego sales based on this sample.

-------------

1.  Describe what you see in the data set (variables, observations, etc)

2.  What are the three most common first names of purchasers?

3.  What are the three most common themes of Lego sets purchased?

4.  Among the most common theme of Lego sets purchased, what is the most common subtheme?

5.  Create data frames for each of the ages in the following categories: "18 and under", "19 - 25", "26 - 35", "36 - 50", "51 and over". HINT - use masks and create separate data frames -OR- create a new column with these categorical labels (more advanced).

6.  Which age group has purchased the highest number of Lego sets.

7.  Which age group has spent the most money on Legos?

8.  Which Lego theme has made the most money for Lego? HINT: Simpler than #5, just use a groupby()

9.  Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. HINT: You will need to split the phone number and get just the first three. You decided what to do about reporting the NaNs.

10.  Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

In [97]:
DF.shape

(620, 14)

In [109]:
DF['first_name'].value_counts()

first_name
Jackson       13
Joseph        11
Jacob         11
Michael       10
Kaitlyn        8
              ..
Gunnar         1
David          1
Lawrence       1
John'Dunae     1
Kimberly       1
Name: count, Length: 211, dtype: int64

In [111]:
DF['theme'].value_counts()

theme
Star Wars                  75
Nexo Knights               64
Gear                       55
Mixels                     55
City                       45
Friends                    42
Ninjago                    38
Duplo                      35
Bionicle                   34
Creator                    25
DC Comics Super Heroes     22
Elves                      22
Marvel Super Heroes        19
Dimensions                 18
Disney Princess            15
The Angry Birds Movie      11
Technic                    10
Architecture               10
Minecraft                   9
Advanced Models             4
Ghostbusters                3
Seasonal                    3
Collectable Minifigures     3
Ideas                       2
Classic                     1
Name: count, dtype: int64

In [113]:
DF['subtheme'].value_counts()

subtheme
Skybound                     23
Ultimate                     22
Fire                         22
Series 9                     21
Series 7                     18
                             ..
Digital media                 1
Frozen                        1
Valentine's Day               1
Ultimate Collector Series     1
Miles from Tomorrowland       1
Name: count, Length: 64, dtype: int64

In [189]:
list(DF)

['first_name',
 'last_name',
 'age',
 'phone_number',
 'set_id',
 'number',
 'theme',
 'subtheme',
 'year',
 'name',
 'pieces',
 'us_price',
 'image_url',
 'quantity']

In [179]:
mask = (DF['age'] >= 18)
DF18 = DF[mask]
DF18

first_name  last_name  age  phone_number  set_id   number  \
0     Kimberly  Beckstead   24  216-555-2549   24701    76062   
1         Neel     Garvin   35  819-555-3189   25626    70595   
2         Neel     Garvin   35  819-555-3189   24665    21031   
3      Chelsea   Bouchard   41           NaN   24695    31048   
4      Chelsea   Bouchard   41           NaN   25626    70595   
..         ...        ...  ...           ...     ...      ...   
610   Jennifer    Reinert   29  402-555-0467   24660    71241   
616    Spencer     Morgan   28  784-555-3455   26041    41580   
617    Spencer     Morgan   28  784-555-3455   26060  5005051   
618     Amelia    Hageman   40  336-555-1950   24702    76063   
619     Amelia    Hageman   40  336-555-1950   24720    10830   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
2              Architecture                   NaN  2018   
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
..                      ...                   ...   ...   
610              Dimensions              Fun Pack  2018   
616                  Mixels              Series 9  2018   
617                    Gear         Digital Media  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                                       name  pieces  us_price  \
0                            Robin vs. Bane    77.0      9.99   
1                      Ultra Stealth Raider  1093.0    119.99   
2                              Burj Khalifa   333.0     39.99   
3                            Lakeside Lodge   368.0     29.99   
4                      Ultra Stealth Raider  1093.0    119.99   
..                                      ...     ...       ...   
610                        Fun Pack: Slimer    33.0     14.99   
616                                    Myke    63.0      4.99   
617  Friends of Heartlake City Girlz 4 Life     NaN     19.99   
618              The Flash vs. Captain Cold    88.0      9.99   
619                           Minnie's Café    27.0     19.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
1    http://images.brickset.com/sets/images/70595-1...         1  
2    http://images.brickset.com/sets/images/21031-1...         1  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
..                                                 ...       ...  
610  http://images.brickset.com/sets/images/71241-1...         3  
616                                                NaN         2  
617                                                NaN         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[603 rows x 14 columns]

In [181]:
mask = (DF['age'] >= 19) & (DF['age'] <= 25)
DF25 = DF[mask]
DF25

first_name    last_name  age  phone_number  set_id number  \
0     Kimberly    Beckstead   24  216-555-2549   24701  76062   
6      Bryanna        Welsh   19           NaN   24797  75138   
7      Bryanna        Welsh   19           NaN   24701  76062   
10       Chase  Fortenberry   19  205-555-3704   24707  10801   
11       Chase  Fortenberry   19  205-555-3704   24713  10809   
..         ...          ...  ...           ...     ...    ...   
591      Paige          Ice   21           NaN   26038  41577   
592      Aiden       Ganley   19  786-555-5067   25639  70324   
593      Aiden       Ganley   19  786-555-5067   24756  70310   
608    Carolyn       Quarry   23  567-555-7649   25627  70596   
609    Carolyn       Quarry   23  567-555-7649   24806  71311   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
6                 Star Wars             Episode V  2018   
7    DC Comics Super Heroes         Mighty Micros  2018   
10                    Duplo                   NaN  2018   
11                    Duplo                   NaN  2018   
..                      ...                   ...   ...   
591                  Mixels              Series 9  2018   
592            Nexo Knights                   NaN  2018   
593            Nexo Knights                   NaN  2018   
608                 Ninjago  Rise of the Villains  2018   
609                Bionicle         Miscellaneous  2018   

                             name  pieces  us_price  \
0                  Robin vs. Bane    77.0      9.99   
6                     Hoth Attack   233.0     24.99   
7                  Robin vs. Bane    77.0      9.99   
10                   Baby Animals    13.0      9.99   
11                  Police Patrol    15.0     14.99   
..                            ...     ...       ...   
591                         Mysto    64.0      4.99   
592          Merlok's Library 2.0   288.0     24.99   
593       Knighton Battle Blaster    76.0      9.99   
608          Samurai X Cave Chaos  1253.0    119.99   
609  Kopaka and Melum - Unity set   171.0     24.99   

                                             image_url  quantity  
0    http://images.brickset.com/sets/images/76062-1...         1  
6    http://images.brickset.com/sets/images/75138-1...         1  
7    http://images.brickset.com/sets/images/76062-1...         3  
10   http://images.brickset.com/sets/images/10801-1...         1  
11   http://images.brickset.com/sets/images/10809-1...         1  
..                                                 ...       ...  
591                                                NaN         1  
592  http://images.brickset.com/sets/images/70324-1...         1  
593  http://images.brickset.com/sets/images/70310-1...         2  
608                                                NaN         2  
609  http://images.brickset.com/sets/images/71311-1...         2  

[129 rows x 14 columns]

In [182]:
mask = (DF['age'] >= 26) & (DF['age'] <= 35)
DF35 = DF[mask]
DF35

first_name  last_name  age  phone_number  set_id   number         theme  \
1         Neel     Garvin   35  819-555-3189   25626    70595       Ninjago   
2         Neel     Garvin   35  819-555-3189   24665    21031  Architecture   
27       Jacob  Nzabanita   31  339-555-2572   24732    41117       Friends   
28       Jacob  Nzabanita   31  339-555-2572   24793    75133     Star Wars   
29       Jacob  Nzabanita   31  339-555-2572   24723    41172         Elves   
..         ...        ...  ...           ...     ...      ...           ...   
606   Benjamin       Park   33           NaN   24957    71309      Bionicle   
607      Miles       Hill   34           NaN   24758    70312  Nexo Knights   
610   Jennifer    Reinert   29  402-555-0467   24660    71241    Dimensions   
616    Spencer     Morgan   28  784-555-3455   26041    41580        Mixels   
617    Spencer     Morgan   28  784-555-3455   26060  5005051          Gear   

                 subtheme  year                                    name  \
1    Rise of the Villains  2018                    Ultra Stealth Raider   
2                     NaN  2018                            Burj Khalifa   
27               Pop Star  2018                      Pop Star TV Studio   
28            Battlefront  2018              Rebel Alliance Battle Pack   
29                    NaN  2018              The Water Dragon Adventure   
..                    ...   ...                                     ...   
606                   Toa  2018                  Onua - Uniter of Earth   
607                   NaN  2018                     Lance's Mecha Horse   
610              Fun Pack  2018                        Fun Pack: Slimer   
616              Series 9  2018                                    Myke   
617         Digital Media  2018  Friends of Heartlake City Girlz 4 Life   

     pieces  us_price                                          image_url  \
1    1093.0    119.99  http://images.brickset.com/sets/images/70595-1...   
2     333.0     39.99  http://images.brickset.com/sets/images/21031-1...   
27    194.0     19.99  http://images.brickset.com/sets/images/41117-1...   
28    101.0     12.99  http://images.brickset.com/sets/images/75133-1...   
29    212.0     19.99  http://images.brickset.com/sets/images/41172-1...   
..      ...       ...                                                ...   
606   143.0     19.99  http://images.brickset.com/sets/images/71309-1...   
607   237.0     19.99  http://images.brickset.com/sets/images/70312-1...   
610    33.0     14.99  http://images.brickset.com/sets/images/71241-1...   
616    63.0      4.99                                                NaN   
617     NaN     19.99                                                NaN   

     quantity  
1           1  
2           1  
27          1  
28          2  
29          1  
..        ...  
606         2  
607         2  
610         3  
616         2  
617         1  

[183 rows x 14 columns]

In [183]:
mask = (DF['age'] >= 36) & (DF['age'] <= 50)
DF50 = DF[mask]
DF50

first_name       last_name  age  phone_number  set_id number  \
3      Chelsea        Bouchard   41           NaN   24695  31048   
4      Chelsea        Bouchard   41           NaN   25626  70595   
5      Chelsea        Bouchard   41           NaN   24721  10831   
8        Caleb  Garcia-Wideman   37  907-555-9236   24730  41115   
9        Caleb  Garcia-Wideman   37  907-555-9236   25611  21127   
..         ...             ...  ...           ...     ...    ...   
598     Payton          Milsap   42           NaN   24702  76063   
599     Payton          Milsap   42           NaN   24678  60109   
602  Stephanie        Harrison   42           NaN   24736  41121   
618     Amelia         Hageman   40  336-555-1950   24702  76063   
619     Amelia         Hageman   40  336-555-1950   24720  10830   

                      theme              subtheme  year  \
3                   Creator                   NaN  2018   
4                   Ninjago  Rise of the Villains  2018   
5                     Duplo                   NaN  2018   
8                   Friends                   NaN  2018   
9                 Minecraft         Minifig-scale  2018   
..                      ...                   ...   ...   
598  DC Comics Super Heroes         Mighty Micros  2018   
599                    City                  Fire  2018   
602                 Friends        Adventure Camp  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   
619                   Duplo                   NaN  2018   

                           name  pieces  us_price  \
3                Lakeside Lodge   368.0     29.99   
4          Ultra Stealth Raider  1093.0    119.99   
5          My First Caterpillar    19.0      9.99   
8      Emma's Creative Workshop   108.0      9.99   
9                  The Fortress     NaN    109.99   
..                          ...     ...       ...   
598  The Flash vs. Captain Cold    88.0      9.99   
599                   Fire Boat   412.0     79.99   
602      Adventure Camp Rafting   320.0     29.99   
618  The Flash vs. Captain Cold    88.0      9.99   
619               Minnie's Café    27.0     19.99   

                                             image_url  quantity  
3    http://images.brickset.com/sets/images/31048-1...         1  
4    http://images.brickset.com/sets/images/70595-1...         1  
5    http://images.brickset.com/sets/images/10831-1...         1  
8    http://images.brickset.com/sets/images/41115-1...         1  
9    http://images.brickset.com/sets/images/21127-1...         2  
..                                                 ...       ...  
598  http://images.brickset.com/sets/images/76063-1...         3  
599  http://images.brickset.com/sets/images/60109-1...         2  
602  http://images.brickset.com/sets/images/41121-1...         1  
618  http://images.brickset.com/sets/images/76063-1...         2  
619  http://images.brickset.com/sets/images/10830-1...         4  

[216 rows x 14 columns]

In [184]:
mask = (DF['age'] >= 50)
DF51 = DF[mask]
DF51

first_name      last_name  age  phone_number  set_id number       theme  \
31      Hannah  Drews Stunkel   55  339-555-6320   25628  70590     Ninjago   
32      Hannah  Drews Stunkel   55  339-555-6320   25624  70593     Ninjago   
33      Hannah  Drews Stunkel   55  339-555-6320   24734  41119     Friends   
38       Riley            Ott   51  517-555-2093   24785  75126   Star Wars   
39       Riley            Ott   51  517-555-2093   24772  70604     Ninjago   
..         ...            ...  ...           ...     ...    ...         ...   
564      Elias        Garrett   50  832-555-9165   25898  75145   Star Wars   
565      Elias        Garrett   50  832-555-9165   24725  41174       Elves   
594     Dustin      Vanvuuren   51  812-555-1009   24709  10803       Duplo   
595     Dustin      Vanvuuren   51  812-555-1009   24807  71235  Dimensions   
596     Dustin      Vanvuuren   51  812-555-1009   25920  75151   Star Wars   

             subtheme  year                       name  pieces  us_price  \
31                NaN  2018    Airjitzu Battle Grounds   666.0     59.99   
32           Skybound  2018       The Green NRG Dragon   567.0     49.99   
33                NaN  2018     Heartlake Cupcake Cafe   439.0     39.99   
38      MicroFighters  2018    First Order Snowspeeder    91.0      9.99   
39           Skybound  2018         Tiger Widow Island   450.0     49.99   
..                ...   ...                        ...     ...       ...   
564  Original Content  2018            Eclipse Fighter   363.0     29.99   
565               NaN  2018          The Starlight Inn   343.0     29.99   
594               NaN  2018                     Arctic    40.0     29.99   
595        Level Pack  2018  Level Pack: Midway Arcade    96.0     29.99   
596       Episode III  2018           Clone Turbo Tank   903.0    109.99   

                                             image_url  quantity  
31                                                 NaN         1  
32   http://images.brickset.com/sets/images/70593-1...         3  
33   http://images.brickset.com/sets/images/41119-1...         1  
38   http://images.brickset.com/sets/images/75126-1...         2  
39   http://images.brickset.com/sets/images/70604-1...         1  
..                                                 ...       ...  
564  http://images.brickset.com/sets/images/75145-1...         1  
565  http://images.brickset.com/sets/images/41174-1...         1  
594  http://images.brickset.com/sets/images/10803-1...         2  
595  http://images.brickset.com/sets/images/71235-1...         1  
596  http://images.brickset.com/sets/images/75151-1...         1  

[75 rows x 14 columns]

In [ ]:
#  DF18, DF25, DF35, DF50, DF51

In [194]:
DF18['us_price'].sum()

17583.97

In [198]:
DF25['us_price'].sum()

3629.710000000001

In [200]:
DF35['us_price'].sum()

5260.169999999998

In [202]:
DF50['us_price'].sum()

6641.84

In [204]:
DF51['us_price'].sum()

2297.2499999999995

In [243]:
my_columns = ['us_price','theme']
DF[new_columns].groupby('theme').sum()

us_price
theme                            
Advanced Models            679.96
Architecture               389.90
Bionicle                   549.66
City                      1476.55
Classic                     29.99
Collectable Minifigures     11.97
Creator                    409.75
DC Comics Super Heroes     450.78
Dimensions                 369.82
Disney Princess            167.85
Duplo                      854.65
Elves                      809.78
Friends                    924.58
Gear                      1056.45
Ghostbusters               469.97
Ideas                      139.98
Marvel Super Heroes        539.81
Minecraft                 1439.91
Mixels                     274.45
Nexo Knights              1569.36
Ninjago                   1649.62
Seasonal                    29.97
Star Wars                 2842.25
Technic                    492.90
The Angry Birds Movie      375.89

In [257]:
DF['AreaCode'] = DF['phone_number'].str.split('-').str[0]

show(DF)

In [283]:
DF_new = DF.dropna()
inplace = True
DF_new

first_name  last_name  age  phone_number  set_id number  \
0     Kimberly  Beckstead   24  216-555-2549   24701  76062   
1         Neel     Garvin   35  819-555-3189   25626  70595   
12       Kevin       Cruz   20  947-555-7946   24767  70335   
13      Connor      Brown   36  516-555-4310   24548  71230   
14      Connor      Brown   36  516-555-4310   24913  60112   
..         ...        ...  ...           ...     ...    ...   
610   Jennifer    Reinert   29  402-555-0467   24660  71241   
613    Natalia   Vonduyke   16  859-555-3818   24765  70332   
614    Natalia   Vonduyke   16  859-555-3818   25922  75149   
615     Talise   Nieukirk   16  801-555-2343   24902  41556   
618     Amelia    Hageman   40  336-555-1950   24702  76063   

                      theme              subtheme  year  \
0    DC Comics Super Heroes         Mighty Micros  2018   
1                   Ninjago  Rise of the Villains  2018   
12             Nexo Knights              Ultimate  2018   
13               Dimensions              Fun Pack  2018   
14                     City                  Fire  2018   
..                      ...                   ...   ...   
610              Dimensions              Fun Pack  2018   
613            Nexo Knights              Ultimate  2018   
614               Star Wars     The Force Awakens  2018   
615                  Mixels              Series 7  2018   
618  DC Comics Super Heroes         Mighty Micros  2018   

                           name  pieces  us_price  \
0                Robin vs. Bane    77.0      9.99   
1          Ultra Stealth Raider  1093.0    119.99   
12             Ultimate Lavaria    69.0      9.99   
13          Fun Pack: Doc Brown    69.0     14.99   
14                  Fire Engine   376.0     49.99   
..                          ...     ...       ...   
610            Fun Pack: Slimer    33.0     14.99   
613              Ultimate Aaron    82.0      9.99   
614   Resistance X-wing Fighter   740.0     79.99   
615                      Tiketz    62.0      4.99   
618  The Flash vs. Captain Cold    88.0      9.99   

                                             image_url  quantity AreaCode  \
0    http://images.brickset.com/sets/images/76062-1...         1      216   
1    http://images.brickset.com/sets/images/70595-1...         1      819   
12   http://images.brickset.com/sets/images/70335-1...         2      947   
13   http://images.brickset.com/sets/images/71230-1...         1      516   
14   http://images.brickset.com/sets/images/60112-1...         1      516   
..                                                 ...       ...      ...   
610  http://images.brickset.com/sets/images/71241-1...         3      402   
613  http://images.brickset.com/sets/images/70332-1...         1      859   
614  http://images.brickset.com/sets/images/75149-1...         1      859   
615  http://images.brickset.com/sets/images/41556-1...         2      801   
618  http://images.brickset.com/sets/images/76063-1...         2      336   

    areacode  
0        216  
1        819  
12       947  
13       516  
14       516  
..       ...  
610      402  
613      859  
614      859  
615      801  
618      336  

[287 rows x 16 columns]

In [291]:
my_columns = ['areacode','us_price']
DF[my_columns].groupby('areacode').describe()

us_price                                                           
            count       mean        std    min    25%    50%     75%     max
areacode                                                                    
205           7.0  19.990000  10.408330   9.99  12.49  19.99   22.49   39.99
206           1.0  24.990000        NaN  24.99  24.99  24.99   24.99   24.99
209           5.0  65.990000  60.249481  19.99  29.99  49.99   59.99  169.99
210           4.0  44.990000  26.457513  19.99  27.49  39.99   57.49   79.99
212           1.0  24.990000        NaN  24.99  24.99  24.99   24.99   24.99
...           ...        ...        ...    ...    ...    ...     ...     ...
970           3.0  18.323333  12.583057   4.99  12.49  19.99   24.99   29.99
973           6.0  74.156667  53.142889   9.99  31.24  79.99  109.99  139.99
979           2.0  79.990000  42.426407  49.99  64.99  79.99   94.99  109.99
980           2.0   5.490000   0.707107   4.99   5.24   5.49    5.74    5.99
989           2.0  14.990000  14.142136   4.99   9.99  14.99   19.99   24.99

[156 rows x 8 columns]

### 1

- This data frame had 620 observations, 14 variables, with mixed numerical and categorical data.

### 2

- The top 3 names in the data frame are Jackson at 13, and Joseph and Jacob both at 11.

### 3

- Star Wars, Nexo Knights, and Gear are the most popular set themes.

### 4

- The top 3 subthemes are Skybound, Ultimate, and Fire.

### 5

- 

### 6

- The 18 and below category purchased the most sets out of all of the age categories.

### 7

- The 18 and below category spent the most out of all other ages with $17,583.97 total spent.

### 8

- Star Wars themed sets made $2,842.25 total. 

### 9

- 

### 10
 g9 roupby()

Which area code has spent the most money on Legos? In the US the area code is the first 3 digits of a phone number. HINT: You will need to split the phone number and get just the first three. You decided what to do about reporting10  the NaNs.

Come up with a question you want to answer using these data, and write it down. Then, create a data visualization that answers the question, and explain how your visualization answers the question.

## Essay Response

Give a brief report on what you are reading and what you have learned so far. If you chose your own articles please insert a link to the articles. If you chose a book, please give the book title and author.

1. What drew you to the book/articles
2. What have you learned so far.

eg. In the first chapter of the book the author talks about..... or My first article discusses..... or I disagree with what the author is saying so far.... or I really like what I am learning about because....